log/keyvalue_test.go

```diff
@@ -300,3 +300,61 @@ func testKV[T any](t *testing.T, key string, val T, kv log.KeyValue) {
 	assert.False(t, kv.Value.Empty(), "value empty")
 	assert.Equal(t, kv.Value.AsAny(), T(val), "AsAny wrong value")
 }
+
+func TestAllocationLimits(t *testing.T) {
+	const (
+		runs = 5
+		key  = "key"
+	)
+
+	// Assign testing results to external scope so the compiler doesn't
+	// optimize away the testing statements.
+	var (
+		a     any
+		i     int64
+		f     float64
+		b     bool
+		by    []byte
+		s     string
+		slice []log.Value
+		m     []log.KeyValue
+	)
+
+	assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
+		b = log.Bool(key, true).Value.AsBool()
+	}), "Bool.AsBool")
```

Comment:
I think we can extract a helper like

```go
func assertZeroAllocs(t *testing.T, f func(), msgAndArgs ...interface{}) {
	t.Helper()
	assert.Equal(t, 0.0, testing.AllocsPerRun(5, f, msgAndArgs...)
}
```

Reply:
Per https://github.com/open-telemetry/opentelemetry-go/pull/4957#discussion_r1499773283 I suggest doing it as a separate PR so that more people also review it